src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -66,70 +65,183 @@ public void WriteMessage(HubMessage message, Stream output)
+
+                using (var reader = new JsonTextReader(textReader))
                 {
-                    // PERF: Could probably use the JsonTextReader directly for better perf and fewer allocations
-                    var token = JToken.ReadFrom(reader);
+                    reader.ArrayPool = JsonArrayPool<char>.Shared;
 
-                    if (token == null || token.Type != JTokenType.Object)
-                    {
-                        throw new InvalidDataException($"Unexpected JSON Token Type '{token?.Type}'. Expected a JSON Object.");
-                    }
+                    reader.Read();
 
-                    var json = (JObject)token;
+                    // We're always parsing a JSON object
+                    if (reader.TokenType != JsonToken.StartObject)
+                    {
+                        if (reader.TokenType == JsonToken.None)
+                        {
+                            throw new InvalidDataException("Error reading JSON.");
+                        }
 
-                    // Determine the type of the message
-                    var type = JsonUtils.GetRequiredProperty<int>(json, TypePropertyName, JTokenType.Integer);
+                        throw new InvalidDataException($"Unexpected JSON Token Type '{reader.TokenType}'. Expected a JSON Object.");
+                    }
 
-                    switch (type)
+                    do
                     {
-                        case HubProtocolConstants.InvocationMessageType:
-                            return BindInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamInvocationMessageType:
-                            return BindStreamInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamItemMessageType:
-                            return BindStreamItemMessage(json, binder);
-                        case HubProtocolConstants.CompletionMessageType:
-                            return BindCompletionMessage(json, binder);
-                        case HubProtocolConstants.CancelInvocationMessageType:
-                            return BindCancelInvocationMessage(json);
-                        case HubProtocolConstants.PingMessageType:
-                            return PingMessage.Instance;
-                        default:
-                            throw new InvalidDataException($"Unknown message type: {type}");
+                        switch (reader.TokenType)
+                        {
+                            case JsonToken.PropertyName:
+                                // We only care about parsing top level properties
+                                if (reader.Depth == 1)
+                                {
+                                    string memberName = reader.Value.ToString();
+
+                                    if (string.Equals(memberName, TypePropertyName, StringComparison.Ordinal))
```

Comment:
Is this faster than a switch on `memberName`?

Reply:
I can try it